### PR TITLE
Implement token refreshing for GitHub App

### PR DESCRIPTION
**Pull request by pjlast**
As the title says. Implements token refreshing for all GitHub OAuth related requests.

Currently this is only relevant for GitHub App, but the groundwork for normal GitHub OAuth is also there should it become relevant.

## Test plan

Tests updated where applicable

### CODE REVIEW

File: enterprise/internal/authz/github/app.go [hunk 1]
```diff
@@ -33,32 +36,73 @@ func newAppProvider(
...
+		rawConfig, err := svc.Config.Decrypt(context.Background())
+		if err != nil {
+			return nil, errors.Errorf("external service id=%d config error: %s", svc.ID, err)
+		}
+		var c schema.GitHubConnection
+		if err := jsonc.Unmarshal(rawConfig, &c); err != nil {
+			return nil, errors.Errorf("external service id=%d config error: %s", svc.ID, err)
+		}
+
+		appAuther, err := github.NewGitHubAppAuthenticator(appID, pkey)
```
Comment (pjlast) [inline enterprise/internal/authz/github/app.go:51]: This whole appAuther dance can be factored out to a common function

File: enterprise/internal/authz/github/github.go [hunk 2]
```diff
@@ -343,13 +348,25 @@ func (p *Provider) FetchUserPerms(ctx context.Context, account *extsvc.Account,
...
 
-	return p.fetchUserPermsByToken(ctx, extsvc.AccountID(account.AccountID), tok.AccessToken, opts)
+	oauthToken := &auth.OAuthBearerToken{
+		Token:        tok.AccessToken,
+		RefreshToken: tok.RefreshToken,
+		Expiry:       tok.Expiry,
+	}
+
+	if p.db != nil {
+		// Only used if created by newAppProvider
```
[the remark this answers is not in the document]
Comment (pjlast) [reply enterprise/internal/authz/github/github.go:358]: Hmmmmm I started typing something but you made me think. Perhaps we do need a separate Provider

**Review by kopancek — APPROVED** [review]

File: cmd/gitserver/main.go [hunk 3]
```diff
@@ -402,25 +401,38 @@ func editGitHubAppExternalServiceConfigToken(
...
+	appAuther, err := github.NewGitHubAppAuthenticator(appID, pkey)
 	if err != nil {
 		return "", errors.Wrap(err, "new authenticator with GitHub App")
 	}
 
-	client := github.NewV3Client(logger, svc.URN(), apiURL, auther, cli)
+	appClient := github.NewV3Client(
+		logger.Scoped("app", "github client for github app").
+			With(log.String("appID", appID)),
+		svc.URN(), apiURL, appAuther, cli)
```
Comment (mrnugget) [inline cmd/gitserver/main.go:417]: Sweet intendation jesus, what's going on here, did a car drive into this

File: enterprise/cmd/repo-updater/internal/authz/perms_syncer.go [hunk 4]
```diff
@@ -254,7 +254,7 @@ func (s *PermsSyncer) listPrivateRepoNamesBySpecs(ctx context.Context, repoSpecs
 	return repoNames, nil
 }
 
-func (s *PermsSyncer) getUserGitHubAppInstallations(ctx context.Context, acct *extsvc.Account) ([]gh.Installation, error) {
+func (s *PermsSyncer) getUserGitHubAppInstallations(ctx context.Context, acct *extsvc.Account, db database.DB) ([]gh.Installation, error) {
```
Comment (mrnugget) [inline enterprise/cmd/repo-updater/internal/authz/perms_syncer.go:257]: Hello, it's Nit Pick here, you might know me from "There's a trailing whitespace at the end of this line" or "I'd put a newline here" and today I want to ask you: how about making `db` the 2nd parameter here? I think we have a convention of "dependencies" coming first, before the actual args.

File: enterprise/internal/authz/github/app.go [hunk 5]
```diff
@@ -33,32 +35,47 @@ func newAppProvider(
 		return nil, errors.Wrap(err, "decode private key")
 	}
 
-	auther, err := auth.NewOAuthBearerTokenWithGitHubApp(appID, pkey)
-	if err != nil {
-		return nil, errors.Wrap(err, "new authenticator with GitHub App")
+	apiURL, _ := github.APIRoot(baseURL)
+	var installationAuther auth.Authenticator
+	if svc != nil {
```
Comment (mrnugget) [inline enterprise/internal/authz/github/app.go:40]: So just to double-check: if we don't have a `svc` (when does that happen?) then `installationAuther` is nil, right? And everything still works and we don't run into this whole nil-interface-is-not-really-nil thing that Go is infamous for?

File: enterprise/internal/authz/github/app.go [hunk 6]
```diff
@@ -33,32 +35,47 @@ func newAppProvider(
...
+		apiURL, _ := github.APIRoot(baseURL)
+		appClient := github.NewV3Client(
+			log.Scoped("app", "github client for github app").
+				With(log.String("appID", appID)),
+			urn, apiURL, appAuther, cli)
+
+		installationAuther, err = github.NewGitHubAppInstallationAuthenticator(installationID, c.Token, svc.TokenExpiresAt, database.GetAppInstallationRefreshFunc(db.ExternalServices(), installationID, svc, appClient))
+		if err != nil {
+			return nil, errors.Wrap(err, "new GitHub App installation auther")
+		}
```
Comment (mrnugget) [inline enterprise/internal/authz/github/app.go:64]: Can't we pull all of this out into a separate function? `buildInstallationAuther` or something like that? It's hard to follow the flow of the outer function, especially since there's multiple clients being initialised etc.)

File: enterprise/internal/authz/github/app.go [hunk 7]
```diff
@@ -33,32 +35,47 @@ func newAppProvider(
...
+			return nil, errors.Wrap(err, "new authenticator with GitHub App")
+		}
+
+		apiURL, _ := github.APIRoot(baseURL)
+		appClient := github.NewV3Client(
+			log.Scoped("app", "github client for github app").
+				With(log.String("appID", appID)),
+			urn, apiURL, appAuther, cli)
+
+		installationAuther, err = github.NewGitHubAppInstallationAuthenticator(installationID, c.Token, svc.TokenExpiresAt, database.GetAppInstallationRefreshFunc(db.ExternalServices(), installationID, svc, appClient))
```
Comment (mrnugget) [inline enterprise/internal/authz/github/app.go:61]: I lack context, but have to admit that this is confusing to me: we build an `appClient`, use it to build an `installationAuther` and then use that `installationAuther` to build another client?

File: enterprise/internal/authz/github/app_test.go [hunk 8]
```diff
@@ -50,6 +50,14 @@ func TestNewAppProvider(t *testing.T) {
 				}, nil
 			}
 
+			if r.Header.Get("Authorization") != "Bearer app-token" {
+				return &http.Response{
+					Status:     http.StatusText(http.StatusUnauthorized),
+					StatusCode: http.StatusUnauthorized,
+					Body:       io.NopCloser(bytes.NewReader([]byte(`invalid access token`))),
+				}, nil
```
Comment (mrnugget) [inline enterprise/internal/authz/github/app_test.go:58]: Why not use `http.Error` in this test? You know:

```go
http.Error(w, "invalid access token", http.StatusUnauthorized)
```

File: enterprise/internal/authz/github/github.go (hunk 2 above)
Comment (mrnugget) [inline enterprise/internal/authz/github/github.go:358]: Are there any other valid uses of `Provider` that don't come from `newAppProvider`? If no, I'd remove the check.

File: internal/database/github_app_helper.go [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+import (
+	"context"
+
+	"github.com/sourcegraph/sourcegraph/internal/conf"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc/github"
+	"github.com/sourcegraph/sourcegraph/internal/jsonc"
+	"github.com/sourcegraph/sourcegraph/internal/types"
+)
+
+func GetAppInstallationRefreshFunc(externalServiceStore ExternalServiceStore, installationID int64, svc *types.ExternalService, appClient *github.V3Client) func(auther *github.GitHubAppInstallationAuthenticator) error {
```
Comment (mrnugget) [inline internal/database/github_app_helper.go:12]: I'd put GitHub in the name of this function, since we're in the `database` package.

File: internal/database/github_app_helper.go [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+	return func(auther *github.GitHubAppInstallationAuthenticator) error {
```
Comment (mrnugget) [inline internal/database/github_app_helper.go:13]: again: I'd really love for these refresh funcs to take `ctx context.Context` so we don't have to use `context.Background()` everywhere

File: internal/database/github_app_helper.go [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+		token, err := appClient.CreateAppInstallationAccessToken(context.Background(), installationID)
+		if err != nil {
+			return err
+		}
+
+		auther.InstallationAccessToken = token.GetToken()
+		auther.Expiry = token.ExpiresAt
```
Comment (mrnugget) [inline internal/database/github_app_helper.go:20]: Is that okay, to modify the parameter?

File: internal/database/github_app_helper.go [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+
+		rawConfig, err := svc.Config.Decrypt(context.Background())
+		if err != nil {
+			return err
+		}
+
+		rawConfig, err = jsonc.Edit(rawConfig, token.GetToken(), "token")
+		if err != nil {
+			return err
+		}
```
Comment (mrnugget) [inline internal/database/github_app_helper.go:30]: We always update the config when we get a new token? TIL.

**Review by mrnugget — APPROVED** [review]
Still missing lots of context, so only reviewed on a pure code basis. I like the new authenticator interface, but I have to say that the other interfaces that are being used to build the RefreshFuncs are confusing me

File: cmd/gitserver/main.go (hunk 3 above)
Comment (pjlast) [reply cmd/gitserver/main.go:417]: Yeah this is why I think scoped loggers should be created in the function and not in the parameter. I'm going to move it

Comment (pjlast) [reply cmd/gitserver/main.go:417]: Issue is just when you want to add something like `appID` which isn't in the function call

Comment (pjlast) [reply cmd/gitserver/main.go:417]: I'll just move it to above the function

File: enterprise/internal/authz/github/app.go (hunk 5 above)
Comment (pjlast) [reply enterprise/internal/authz/github/app.go:40]: If I recall correctly, on startup, providers are tested with a nil `svc` to see if they're configured correctly. Adding a comment so long, but will just confirm this as well

File: enterprise/internal/authz/github/github.go (hunk 2 above)
Comment (pjlast) [reply enterprise/internal/authz/github/github.go:358]: Yes, normal GitHub OAuth Provider

Comment (mrnugget) [reply enterprise/internal/authz/github/github.go:358]: And in that case it's instantiated without a `db`?

Comment (pjlast) [reply enterprise/internal/authz/github/github.go:358]: Correct

File: enterprise/internal/authz/github/app.go (hunk 7 above)
Comment (pjlast) [reply enterprise/internal/authz/github/app.go]: I separated this out to its own function and added a comment to explain whats happening

File: cmd/gitserver/main.go [hunk 13]
```diff
@@ -402,25 +401,36 @@ func editGitHubAppExternalServiceConfigToken(
...
 
 	// NOTE: Use `json.Marshal` breaks the actual external service config that fails
 	// validation with missing "repos" property when no repository has been selected,
 	// due to generated JSON tag of ",omitempty".
-	config, err := jsonc.Edit(rawConfig, token, "token")
+	config, err := jsonc.Edit(rawConfig, token.Token, "token")
 	if err != nil {
 		return "", errors.Wrap(err, "edit token")
 	}
+	externalServiceStore.Update(ctx, conf.Get().AuthProviders, svc.ID,
```
Comment (mrnugget) [inline cmd/gitserver/main.go:429]: I think we should check the `err` returned by this method

(edit: or change last two statements to `return config, externalServiceStore.Update(ctx, ...)`)

File: enterprise/cmd/repo-updater/internal/authz/perms_syncer.go [hunk 14]
```diff
@@ -254,7 +254,7 @@ func (s *PermsSyncer) listPrivateRepoNamesBySpecs(ctx context.Context, repoSpecs
 	return repoNames, nil
 }
 
-func (s *PermsSyncer) getUserGitHubAppInstallations(ctx context.Context, acct *extsvc.Account) ([]gh.Installation, error) {
+func (s *PermsSyncer) getUserGitHubAppInstallations(ctx context.Context, db database.DB, acct *extsvc.Account) ([]gh.Installation, error) {
```
Comment (mrnugget) [inline enterprise/cmd/repo-updater/internal/authz/perms_syncer.go:257]: It's a bit confusing that we now have `db` as a param when there's also `s.db`.

File: enterprise/internal/authz/github/github.go [hunk 15]
```diff
@@ -343,13 +348,25 @@ func (p *Provider) FetchUserPerms(ctx context.Context, account *extsvc.Account,
...
+	}
+
+	return p.fetchUserPermsByToken(ctx, extsvc.AccountID(account.AccountID), oauthToken, opts)
 }
 
 // FetchUserPermsByToken is the same as FetchUserPerms, but it only requires a
 // token.
 func (p *Provider) FetchUserPermsByToken(ctx context.Context, token string, opts authz.FetchPermsOptions) (*authz.ExternalUserPermissions, error) {
-	return p.fetchUserPermsByToken(ctx, "", token, opts)
+	return p.fetchUserPermsByToken(ctx, "", &auth.OAuthBearerToken{Token: token}, opts)
```
Comment (mrnugget) [inline enterprise/internal/authz/github/github.go:369]: Why doesn't this require a full `OAuthBearerToken` with `RefreshFunc` set and all that?

**Review by mrnugget — COMMENTED** [review]
(Did another pass but I'll defer to the rest of the IAM team for reviews here)

File: enterprise/cmd/repo-updater/internal/authz/perms_syncer.go (hunk 14 above)
Comment (pjlast) [reply enterprise/cmd/repo-updater/internal/authz/perms_syncer.go:257]: Thanks for pointing out

File: enterprise/internal/authz/github/github.go (hunk 15 above)
Comment (miveronese) [reply enterprise/internal/authz/github/github.go:369]: Was looking into `fetchUserPermsByToken` just now and stumbled upon these comments... :)
`FetchUserPermsByToken` has been only implemented by GL ([sudo](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/internal/authz/gitlab/sudo.go?L218) and [oauth](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/internal/authz/gitlab/oauth.go?subtree=true)) and [GH](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/internal/authz/github/github.go?L351&subtree=true).

It's used only by the `perms_syncer` [when we fetch the list of private repos ](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/cmd/repo-updater/internal/authz/perms_syncer.go?L658&subtree=true)a user can access in a given code host/extsvc. And the token passed to the `FetchUserPermsByToken` function is the external service token.

Here especially, at first, it looks like adding a full `OAuthBearerToken` with a `RefreshFunc ` will work (we would have to change the interface a little bit and pass to it the externalServiceID instead of an accountID, as this is an external service token).
But — and this is a source of confusion right now to me: when looking at our DB and testing this, I cannot see any refresh tokens stored in the external services, only access tokens.

Comment (miveronese) [reply enterprise/internal/authz/github/github.go:369]: Oops, nvm: the last part (not seeing the refresh token in the DG, in external_services table, config column) is because we don't have, in GH schema, the fields `token.oauth.refresh` , `token.oauth.expiry`, etc, as we have for GL. But this can be solved (it was added in a different PR that ended up replaced)

Comment (pjlast) [reply enterprise/internal/authz/github/github.go:369]: So if it's only used by `external_services` tokens, then this should only be Personal Access Tokens, as we don't support adding other kinds of tokens to External Services at the moment. So if we're able to get the Personal Access Tokens refactor PR merged, then this can be switched to that.

But otherwise it should be fine for now to only be OAuthBearerTokens without any other details.

I know `token.oauth.refresh` and `token.oauth.expiry` was added for GitLab, but I don't think that should have happened. For one, we don't have any way for customers to actually set up an external service using OAuth tokens, only Personal Access Tokens, and also, the `external_services` table already has a `token_expires_at` column, so the `token.oauth.expiry` field shouldn't be necessary

So this is why you won't see any refresh tokens in the external services table. We only support personal access tokens, and they cannot be refreshed.

Comment (pjlast) [reply enterprise/internal/authz/github/github.go:369]: However, that being said, fetching user permissions using an external service entry should be a no-no, so perhaps it's better to remove this part of the interface completely
